src/Commands/core/ArchiveDumpCommands.php

```diff
@@ -0,0 +1,615 @@
+     *
+     * @command archive:dump
+     * @aliases ard
+     *
+     * @option code Archive codebase.
+     * @option files Archive Drupal files.
+     * @option db Archive database SQL dump.
+     * @option destination The full path and filename in which the archive should be stored. If omitted, it will be saved to the drush-backups directory.
+     * @option overwrite Overwrite destination file if exists.
+     * @option description Describe the archive contents.
+     * @option tags Add tags to the archive manifest. Delimit several by commas.
+     * @option overwrite Do not fail if the destination file exists; overwrite it instead. Default is --no-overwrite.
+     * @option generator The generator name to store in the MANIFEST.yml file. The default is "Drush archive-dump".
+     * @option generatorversion The generator version number to store in the MANIFEST file. The default is Drush version.
+     * @option exclude-code-paths Comma-separated list of paths to exclude from the code archive.
+     *
+     * @optionset_sql
+     * @optionset_table_selection
+     *
+     * @bootstrap max configuration
+     *
+     * @param array $options
+     *
+     * @return string
+     *
+     * @throws \Exception
+     */
+    public function dump(array $options = [
+        'code' => false,
+        'files' => false,
+        'db' => false,
+        'destination' => null,
+        'overwrite' => false,
+        'description' => null,
+        'tags' => null,
+        'generator' => null,
+        'generatorversion' => null,
+        'exclude-code-paths' => null,
+    ]): string
+    {
+        $this->prepareArchiveDir();
+
+        if (!$options['code'] && !$options['files'] && !$options['db']) {
+            $options['code'] = $options['files'] = $options['db'] = true;
+        }
+
+        $components = [];
+
+        if ($options['code']) {
+            $components[] = [
+                'name' => self::COMPONENT_CODE,
+                'path' => $this->getCodeComponentPath($options),
+            ];
+        }
+
+        if ($options['files']) {
+            $components[] = [
+                'name' => self::COMPONENT_FILES,
+                'path' => $this->getDrupalFilesComponentPath(),
+            ];
+        }
+
+        if ($options['db']) {
+            $components[] = [
+                'name' => self::COMPONENT_DATABASE,
+                'path' => $this->getDatabaseComponentPath($options),
+            ];
+        }
+
+        return $this->createArchiveFile($components, $options);
+    }
+
+    /**
+     * Creates a temporary directory for the archive.
+     *
+     * @throws \Exception
+     */
+    protected function prepareArchiveDir(): void
+    {
+        $this->filesystem = new Filesystem();
+        $this->archiveDir = Path::join(FsUtils::prepareBackupDir(self::ARCHIVES_DIR_NAME), self::ARCHIVE_SUBDIR_NAME);
+        $this->filesystem->mkdir($this->archiveDir);
+        register_shutdown_function([$this, 'cleanUp']);
```

Comment:
@weitzman: Drush 8 had a number of convenience methods related to temporary files and directories which were automatically cleaned up in Drush's shutdown handler. These were primarily used by make and pm commands, which are no longer part of Drush. Config import and rsync used to use them in Drush 8, but do not in Drush 11. What are your thoughts on direct use of `register_shutdown_function()` for cleanup in light of the lack of this service in Drush 11? Is this the best practice for commands that want cleanup on termination?